File: app/schemas/llm_usage.py
"""LLM 用量统计响应模型（Tech Design M4 §11/§12，Phase 2）。"""
import uuid
from datetime import date, datetime

from pydantic import BaseModel


class ProviderStat(BaseModel):
    provider: str
    calls: int
    success_calls: int
    success_rate: float
    fallback_calls: int  # fallback_index > 0：作为降级目标被调用的次数
    fallback_rate: float
    input_tokens: int
    output_tokens: int
    total_tokens: int


class TaskStat(BaseModel):
    task_type: str
    calls: int
    success_calls: int
    success_rate: float
    input_tokens: int
    output_tokens: int
    total_tokens: int
    avg_latency_ms: int | None = None


class GrantStatus(BaseModel):
    """单个资源包的应用侧估算状态（预警用，控制台余额是最终真值）。"""

    name: str  # "glm_air" | "glm_general"
    providers: list[str]  # 计入该资源包的 provider 名单
    grant_total_tokens: int
    tracked_total_tokens: int
    estimated_remaining_tokens: int
    usage_ratio: float
    usage_level: str | None = None  # §12.2：observe/evaluate_routing/high_value_only/exhausted
    avg_tokens_per_meeting: int | None = None
    estimated_remaining_meetings: int | None = None
    avg_daily_tokens_7d: int | None = None
    estimated_exhaustion_date: date | None = None
    expires_at: datetime
    days_until_expiry: int
    expiry_warning: str | None = None  # §12.1：到期前 30/14/7/1 天提醒文案


class FailureOut(BaseModel):
    created_at: datetime
    meeting_id: uuid.UUID | None = None
    task_type: str
    provider: str
    model: str
    fallback_index: int
    error_type: str | None = None
    error_message: str | None = None


class UsageStatsOut(BaseModel):
    total_calls: int
    total_tokens: int
    by_provider: list[ProviderStat]
    by_task: list[TaskStat]
    grants: list[GrantStatus]
    recent_failures: list[FailureOut]

File: web/src/api/llmUsage.ts
import { request } from "./client";

export interface ProviderStat {
  provider: string;
  calls: number;
  success_calls: number;
  success_rate: number;
  fallback_calls: number;
  fallback_rate: number;
  input_tokens: number;
  output_tokens: number;
  total_tokens: number;
}

export interface TaskStat {
  task_type: string;
  calls: number;
  success_calls: number;
  success_rate: number;
  input_tokens: number;
  output_tokens: number;
  total_tokens: number;
  avg_latency_ms: number | null;
}

export interface GrantStatus {
  name: string;
  providers: string[];
  grant_total_tokens: number;
  tracked_total_tokens: number;
  estimated_remaining_tokens: number;
  usage_ratio: number;
  usage_level: string | null;
  avg_tokens_per_meeting: number | null;
  estimated_remaining_meetings: number | null;
  avg_daily_tokens_7d: number | null;
  estimated_exhaustion_date: string | null;
  expires_at: string;
  days_until_expiry: number;
  expiry_warning: string | null;
  soft_limit_tokens: number | null;
  soft_limit_reached: boolean;
  hard_limit_reached: boolean;
  expired: boolean;
  allow_paid_after_grant: boolean;
  enforcement: string; // none | high_value_only | blocked
  enforcement_reason: string | null; // exhausted | expired
}

export interface UsageFailure {
  created_at: string;
  meeting_id: string | null;
  task_type: string;
  provider: string;
  model: string;
  fallback_index: number;
  error_type: string | null;
  error_message: string | null;
}

export interface UsageStats {
  total_calls: number;
  total_tokens: number;
  by_provider: ProviderStat[];
  by_task: TaskStat[];
  grants: GrantStatus[];
  recent_failures: UsageFailure[];
}

export function getUsageStats(): Promise<UsageStats> {
  return request<UsageStats>("/api/llm-usage/stats");
}
